File: apps/api/src/openfga/write-error.ts
import { FgaApiError, FgaApiInternalError, FgaApiRateLimitExceededError } from "@openfga/sdk";

// OpenFGA write/delete 오류 분류(lazyfga-15 hardening). idp(웹훅 동기화)·permission(lazyfga-20
// grant/revoke)이 공유한다. 분류 의미는 두 곳에서 동일해야 하므로 단일 모듈로 둔다.

const TRANSIENT_CODES = new Set([
  "ECONNREFUSED",
  "ENOTFOUND",
  "ECONNRESET",
  "ETIMEDOUT",
  "EAI_AGAIN",
]);

/**
 * 일시적(재시도 가능) OpenFGA 오류인가. **HTTP status 기준**:
 * - 5xx/429, 또는 HTTP 응답이 없음(statusCode undefined = 네트워크 단절) → transient.
 *   ECONNRESET/소켓 끊김 등은 SDK가 .code를 보존하지 않으므로 statusCode 부재로 잡는다(FgaApiError 포함).
 * - 4xx(결정적): **메시지 free-text로는 절대 transient 판정하지 않는다**(이벤트 값이 'timeout' 등을
 *   품어 무한재시도 유발하는 것 방지).
 * write/delete/read가 공유한다(transient → 502, 그 외 → 결정적 처리).
 */
export function isTransientApiError(e: unknown): boolean {
  if (e instanceof FgaApiInternalError || e instanceof FgaApiRateLimitExceededError) return true;
  const statusCode = (e as { statusCode?: number } | null)?.statusCode;
  if (typeof statusCode === "number" && (statusCode >= 500 || statusCode === 429)) return true;
  if (statusCode === undefined) {
    if (e instanceof FgaApiError) return true; // HTTP 응답 없음 = 네트워크 단계 오류.
    const code = (e as { code?: string } | null)?.code;
    if (code && TRANSIENT_CODES.has(code)) return true;
    const m = String((e as { message?: string } | null)?.message ?? e).toLowerCase();
    if (
      m.includes("fetch failed") ||
      m.includes("network") ||
      m.includes("timeout") ||
      m.includes("econnrefused")
    )
      return true;
    // 정체불명 + status 없음 → 무한재시도 방지 위해 결정적으로 취급(false).
  }
  return false;
}

/**
 * write/delete 오류 분류(lazyfga-15 hardening). transient면 502, 아니면 멱등 흡수는
 * invalid-input 코드 + op별 **정확한** 패턴이 함께 맞을 때만 한다:
 * - write 멱등 = 이미 존재하는 tuple(duplicate): "already exists" | "duplicate".
 * - delete 멱등 = 없는 tuple(missing): "cannot delete" | "does not exist".
 * delete 패턴에서 막연한 "not found"는 제외한다 — invalid-input + "relation/type not found" 같은
 * **실제** 거부를 missing no-op으로 삼켜 deleted:false로 숨기는 over-match를 막는다(LFGA-20 review).
 */
export function classifyWriteError(
  e: unknown,
  op: "write" | "delete",
): { idempotent: boolean; transient: boolean } {
  if (isTransientApiError(e)) return { idempotent: false, transient: true };
  // 결정적(4xx 등): 멱등 흡수는 invalid-input 코드/신호 + op별 정확 패턴이 함께 맞을 때만.
  const apiCode = (e as { responseData?: { code?: string } } | null)?.responseData?.code;
  const msg = String((e as { message?: string } | null)?.message ?? e).toLowerCase();
  const isInvalidInput =
    apiCode === "write_failed_due_to_invalid_input" ||
    msg.includes("write_failed_due_to_invalid_input");
  const opPattern =
    op === "write"
      ? msg.includes("already exists") || msg.includes("duplicate")
      : msg.includes("cannot delete") || msg.includes("does not exist");
  return { idempotent: isInvalidInput && opPattern, transient: false };
}

File: apps/api-go/internal/openfga/writeerror/writeerror.go
// Package writeerror는 OpenFGA write/delete 오류를 분류한다(TS write-error.ts 포팅).
// idp(웹훅 동기화)·permission(grant/revoke)이 공유하므로 분류 의미는 단일 소스로 둔다.
package writeerror

import (
	"errors"
	"net"
	"strings"

	fga "github.com/openfga/go-sdk"
)

// Op는 멱등 흡수 패턴을 op별로 구분하기 위한 연산 종류다.
type Op string

const (
	OpWrite  Op = "write"
	OpDelete Op = "delete"
)

// Classification은 결정적/일시적/멱등 판정 결과다.
type Classification struct {
	Idempotent bool
	Transient  bool
}

// invalidInputCode는 OpenFGA가 중복 write / 없는 delete에 대해 내는 코드다.
const invalidInputCode = "write_failed_due_to_invalid_input"

// responseStatusCoder는 SDK Fga*Error가 공통으로 구현하는 상태코드 접근자다.
type responseStatusCoder interface {
	ResponseStatusCode() int
}

// IsTransientAPIError는 재시도 가능한(일시적) OpenFGA/SDK 오류인지 보고한다.
//   - HTTP status >=500 또는 429 → transient.
//   - status가 없음(네트워크 단계): SDK 네트워크 오류(*url.Error/net.Error), 인식되는
//     네트워크 오류 클래스, 또는 parity 메시지 패턴이면 transient.
//   - 그 외 status 없는 오류는 결정적(정체불명 + status 없음 → 무한재시도 방지, false).
//   - 결정적 4xx는 메시지 내용과 무관하게 절대 transient가 아니다.
func IsTransientAPIError(e error) bool {
	if e == nil {
		return false
	}
	// FgaApiInternalError(5xx)·RateLimit(429)은 항상 상태코드가 규칙을 만족하지만
	// 명시적으로도 잡아 TS의 instanceof 분기와 1:1 대응시킨다.
	var internalErr fga.FgaApiInternalError
	if errors.As(e, &internalErr) {
		return true
	}
	var rateErr fga.FgaApiRateLimitExceededError
	if errors.As(e, &rateErr) {
		return true
	}

	// HTTP status가 있는 오류: 5xx/429만 transient, 4xx는 결정적.
	var coder responseStatusCoder
	if errors.As(e, &coder) {
		code := coder.ResponseStatusCode()
		return code >= 500 || code == 429
	}

	// 여기부터는 HTTP status가 없는 오류.
	if isNetworkError(e) {
		return true
	}
	return matchesTransientMessage(e)
}

// ClassifyWriteError는 write/delete 오류를 {Idempotent, Transient}로 분류한다.
// transient면 그대로, 아니면 invalid-input 신호 + op별 정확 패턴이 함께 맞을 때만 멱등 흡수한다.
//   - write 멱등 = 중복 tuple: "already exists" | "duplicate".
//   - delete 멱등 = 없는 tuple: "cannot delete" | "does not exist".
//
// delete에서 막연한 "not found"는 제외한다 — 실제 거부(type/relation not found)를
// missing no-op으로 삼켜 숨기는 over-match를 막는다(LFGA-20 review).
func ClassifyWriteError(e error, op Op) Classification {
	if IsTransientAPIError(e) {
		return Classification{Idempotent: false, Transient: true}
	}
	msg := strings.ToLower(errMessage(e))
	isInvalidInput := responseCode(e) == invalidInputCode || strings.Contains(msg, invalidInputCode)

	var opPattern bool
	switch op {
	case OpWrite:
		opPattern = strings.Contains(msg, "already exists") || strings.Contains(msg, "duplicate")
	case OpDelete:
		opPattern = strings.Contains(msg, "cannot delete") || strings.Contains(msg, "does not exist")
	}
	return Classification{Idempotent: isInvalidInput && opPattern, Transient: false}
}

// responseCode는 검증 오류(FgaApiValidationError)의 API 코드 문자열을 반환한다.
// TS의 `e.responseData?.code` 경로에 대응한다.
func responseCode(e error) string {
	var v fga.FgaApiValidationError
	if errors.As(e, &v) {
		return string(v.ResponseCode())
	}
	return ""
}

// errMessage는 오류 메시지를 안전하게 추출한다.
func errMessage(e error) string {
	if e == nil {
		return ""
	}
	return e.Error()
}

// isNetworkError는 네트워크 단계 오류를 판별한다. net.Error는 *url.Error·*net.OpError·
// *net.DNSError·syscall.Errno를 모두 포함한다(전부 Timeout()/Temporary()를 구현). SDK는
// 연결 실패(연결 거부/리셋/타임아웃/DNS)를 *url.Error로 감싸므로 net.Error 하나로 잡힌다.
// TS의 "FgaApiError(status 부재) + TRANSIENT_CODES → transient"에 대응한다.
func isNetworkError(e error) bool {
	var netErr net.Error
	return errors.As(e, &netErr)
}

// matchesTransientMessage는 TS parity 메시지 패턴 폴백이다.
func matchesTransientMessage(e error) bool {
	m := strings.ToLower(errMessage(e))
	return strings.Contains(m, "fetch failed") ||
		strings.Contains(m, "network") ||
		strings.Contains(m, "timeout") ||
		strings.Contains(m, "econnrefused")
}
